File: c#_tools/NeuralNetworkFromScratch/NeuralNetworkLayer.cs


/// <summary>
/// Represents one layer in a neural network
/// </summary>
public class NeuralNetworkLayer
{
    private readonly float[] inputs;
    private readonly float[,] lstOfWeights;
    private readonly float[,] biases;

    public NeuralNetworkLayer(int numInputs, int numNeurons)
    {
        this.lstOfWeights = MatrixMath.GenerateRandomMatrix((numInputs, numNeurons));
        this.biases = new float[1, numNeurons];

        if (lstOfWeights.GetLength(0) != biases.Length)
        {
            throw new Exception("the number of sets of weights and the number of biases should match");
        }
    }

    public float[,] Forward(float[,] inputs)
    {
        var dotProduct = MatrixMath.MultiplyMatrices(inputs, lstOfWeights);

        for (int i = 0; i < dotProduct.GetLength(0); i++)
        {
            for (int j = 0; j < dotProduct.GetLength(1); j++)
            {
                dotProduct[i, j] += biases[0,j];
            }
        }

        return dotProduct;
    }

    // TODO: modify to perform matrix multiplication (i.e. multiple sets of inputs times multiple sets of weights)
    // public List<float> ComputeLayerOutput()
    // {
    //     List<float> layerOutputs = new(biases.Length);
    //     for (int i = 0; i < biases.Length; i++)
    //     {
    //         float output = 0;

    //         // grab the ith array of weights
    //         var weights = Enumerable.Range(0, lstOfWeights.GetLength(1))
    //         .Select(j => lstOfWeights[i, j]).ToArray();

    //         // multiply weight and input values and add them
    //         foreach (var (weight, input) in weights.Zip(inputs, (weight, input) => (weight, input)))
    //         {
    //             output += weight * input;
    //         }

    //         // finally add the bias
    //         output += biases[i];
    //         layerOutputs.Add(output);
    //     }

    //     return layerOutputs;

    // }
}

File: js_tools/NeuralNetworkFromScratch/src/classes/neural_network_layer.ts
import { Matrix } from "../utils/matrix.js";

export class NeuralNetworkLayer
{
  private readonly lstOfWeights: Matrix;
  private readonly biases: Matrix;

  constructor(numInputs: number, numNeurons: number)
  {
    this.lstOfWeights = new Matrix(numInputs, numNeurons).randomize();
    this.biases = new Matrix(1, numNeurons);

    // console.log(MatrixMath.getShape(this.lstOfWeights), this.lstOfWeights);
    // console.log(MatrixMath.getShape(this.biases), this.biases);
    // console.log(numNeurons);
  }

  public forward(inputs: Matrix): Matrix | undefined
  {
    console.log(inputs.getShape(), inputs);
    console.log(this.lstOfWeights.getShape(), this.lstOfWeights);

    const dotProduct = Matrix.multiply(inputs, this.lstOfWeights);
    
    if (!dotProduct)
    {
      return undefined
    }

    // Add biases to dotProduct (row-wise broadcast)
    dotProduct.sumBias(this.biases.data[0]);

    return dotProduct;
  }

}
